src/visualization/annotation/annotation_base.cpp

```diff
@@ -58,11 +58,32 @@ void AnnotationBase::_sendProgress(double value) {
 
 std::shared_ptr<unity_sframe>
 AnnotationBase::returnAnnotations(bool drop_null) {
-  this->cast_annotations();
-
-  std::shared_ptr<unity_sframe> copy_data =
-      std::static_pointer_cast<unity_sframe>(
+  cast_annotations();
+
+  std::shared_ptr<unity_sframe> copy_data;
+  {
+    // append undefined values
+    if (m_data_na && m_data_na->size()) {
+      auto cast_type = m_data->select_column(m_annotation_column)->dtype();
+      auto na_type = m_data_na->select_column(m_annotation_column)->dtype();
+      if (cast_type != na_type) {
+        auto data_sarray = std::static_pointer_cast<unity_sarray>(
+            m_data_na->select_column(m_annotation_column));
+        auto integer_annotations =
+            data_sarray->astype(cast_type, true);
+        m_data_na->remove_column(m_data_na->column_index(m_annotation_column));
+        m_data_na->add_column(integer_annotations, m_annotation_column);
+      }
+      copy_data =
+          std::static_pointer_cast<unity_sframe>(m_data->append(m_data_na));
+    } else {
+      copy_data = std::static_pointer_cast<unity_sframe>(
           m_data->copy_range(0, 1, m_data->size()));
+    }
+  }
+  // temporary solution
```

Comment:
What is temporary about this solution? What problem is it solving? Why is the solution temporary?

Or if this comment is outdated, please remove it.

Reply:
Haha, I forget to change its position.

Well, the temp solution is to sort by `__idx` column to restore the original order, which is not that efficient. But in the future, if we hit any perf issue, we should come up with a solution that we don't have to split the image set into defined part and undefined part. We can do all operations inline. Since the design by @abhishekpratapa requires all images feeding into neural network to be valid, we come up with this **temp** solution.